microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaTextDocuments.java

```diff
@@ -122,12 +127,16 @@ public void setProjectURI(String projectURI) {
 		 *         project.
 		 */
 		public <T> CompletableFuture<T> executeIfInMicroProfileProject(
-				Function<ProjectLabelInfoEntry, CompletableFuture<T>> code, T defaultValue) {
-			return getProjectInfo(this).thenComposeAsync(projectInfo -> {
-				if (!isMicroProfileProject(projectInfo)) {
+				BiFunction<ProjectLabelInfoEntry, CancelChecker, CompletableFuture<T>> code, T defaultValue) {
+			return CompletableFutures.computeAsync((cancelChecker) -> {
+				ProjectLabelInfoEntry projectInfo = getProjectInfo(this).getNow(null);
+				return Tuple.two(projectInfo, cancelChecker);
+			}).thenCompose((tuple) -> {
+				tuple.getSecond().checkCanceled();
+				if (tuple.getFirst() == null || !isMicroProfileProject(tuple.getFirst())) {

```

Comment:
To have a better read of the code, I suggest that you assign variables at first  like this:

```java
ProjectLabelInfoEntry projectInfo  = tuple.getFirst();
CancelChecker cancelChecker = tuple.getSecond();
```

and you use it after.